### PR TITLE
[XNNPACK] Support running in multi-thread with seperate pthreadpool

### PR DESCRIPTION
**Pull request by wejoncy**
**Description**: Describe your changes.
XNNPACK takes pthreadpool as its internal threadpool implemtation, it couples calculation and parallelization. Thus it's impossible to leverage ORT's threadpool (EIGEN/OPENMP based).  So we enabled pthreadpool  in XNNPACK EP in this PR.

Case 1:  Pthreadpool coexist with ORT-threadpool simply
Expriments setup
hardware:RedMi8A with 8 cores, ARMv7
The two threadpool has the same pool size form 1 to 8.
Two models: mobilenet_v2 and mobilenet_egetppu.
we can see the picture below and draw a conclusion, latency are even higher from 5 threads or more.

image

Case 2:
For the reason of performance regression with 5 more threads, ORT-threads are spinning on CPU and diddn't realease it after computation finished.  It's  equivalent of creating 5x2 threads for parallelization while we have only 8 cpu cores.
So I mannuly disabled spinning after ort-threadpool finished and enabled it when enter ort-threadpool.
The result is quite normal now.
image

Case 3:
Even we achieved a reasonable results with  disabling spinning, Will ORT-threadpool still impact performance of pthreadpool?
we have expriment setting up as: Setting ORT-threadpool size (intra_thread_num) as 1, and only pthreadpool created.
Attention that,  almost a third of ops are running by CPU EP. we are surprisingly find that disabling ort-threadpool is even better in performance than creating two threadpool.

image

Case 4:
Use a unified threadpool between CPU ep and XNNPACK ep.
It's the fastest among all. But if we take the similar workload partition strategy as ORT-threadpool, it could be faster.

image

**Motivation and Context**
- Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h [hunk 1]
```diff
@@ -70,6 +70,10 @@ static const char* const kOrtSessionOptionsUseDeviceAllocatorForInitializers = "
 // "1": default, thread will spin a number of times before blocking
 static const char* const kOrtSessionOptionsConfigAllowInterOpSpinning = "session.inter_op.allow_spinning";
 static const char* const kOrtSessionOptionsConfigAllowIntraOpSpinning = "session.intra_op.allow_spinning";
+// Decide to Create default thread pool or not, if not, the thread pool size is always 1
+// The scenario is for some EPs which has its own threadpool, we don't want two more threadpool coexist
+// XNNPACK is one of the EPs, pthreadpool is used in XNNPACK
+static const char* const kOrtSessionOptionsConfigDisableThreadPool = "session.intra_op.disable_thread_pool";
```
Comment (pranavsharma) [inline include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h:76]: I don't think this config is required. Setting intro_op_num_threads to 1 should suffice as ORTs threadpool uses the calling thread as one of the threads in its pool.

Comment (wejoncy) [reply include/onnxruntime/core/session/onnxruntime_session_options_config_keys.h:76]: Indeed, what this config doing to is set intro_op_num_threads to 1 in `inference_sesstion.cc`.
But without this config, it's hard to prevent ort-devs/users setting intro_op_num_threads  to other numbers.

The motivation behind is to let dev/users config ort-threadpoo/pthreadpool more easily.

File: onnxruntime/core/session/provider_registration.cc [hunk 2]
```diff
@@ -74,7 +75,15 @@ ORT_API_STATUS_IMPL(OrtApis::SessionOptionsAppendExecutionProvider,
...
 #if defined(USE_XNNPACK)
+    // If Xnnpack is enabled and user don't give a specific value, we need to set it same as cpu threadpool size
+    if (provider_options.find("thread_num") == provider_options.end()) {
+      provider_options["thread_num"] = std::to_string(options->value.intra_op_param.thread_pool_size);
+    }
     options->provider_factories.push_back(XnnpackProviderFactoryCreator::Create(provider_options));
+    // XNNPACK owns its own threadpool, so we have to disable ort-threadpool's spinning feature
+    // otherwise it will cause contention.
+    status = onnxruntime::ToOrtStatus(options->value.config_options.AddConfigEntry(
+        kOrtSessionOptionsConfigAllowIntraOpSpinning, "0"));
```
Comment (RandySheriffH) [inline onnxruntime/core/session/provider_registration.cc:86]: Then want if user needs ort threadpool and its spinning for heavy ops on cpu EP?

Comment (pranavsharma) [reply onnxruntime/core/session/provider_registration.cc:86]: I believe this is for mobile scenarios where spinning eats up all the battery.

**Comment by RandySheriffH** [comment]
The approach stratifies the threading between core ort and xnn EP - that leaves us a solid question, what if there are cpu ops expecting multi-threading? The choice we made here indeed brings implementation simplicity, while actually we could go extra steps to make pthread pool also visible to cpu ops to maintain parallelism.

**Comment by yufenglee** [comment]
We ever had 2 threadpool: openmp and eigen-based threadpool, which can lead to tricky issues if both threadpools are used. Will pthreadpool and our current threadpool bring similar issues?

File: onnxruntime/core/providers/xnnpack/nn/softmax.h [hunk 3]
```diff
@@ -3,15 +3,15 @@
 
 #pragma once
```
Comment (RandySheriffH) [inline onnxruntime/core/providers/xnnpack/nn/softmax.h:4]: We do not need headers for these files?

// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License."

**Comment by pranavsharma** [comment]
> We ever had 2 threadpool: openmp and eigen-based threadpool, which can lead to tricky issues if both threadpools are used. Will pthreadpool and our current threadpool bring similar issues?

Ideally you want to have a single threadpool between ORT core and EP. However, currently we don't have a way to pass the threadpool in either direction. Hence, in this case we can simply set the intra_op_num_threads to 1 (which essentially disables the ORT threadpool as the main thread is also used as a thread in the ORT tpool) and ensure most of the heavy ops are executed by xnnpack. This is the fastest way.

File: onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.h [hunk 4]
```diff
@@ -8,17 +8,23 @@
 #include "core/framework/execution_provider.h"
```
Comment (RandySheriffH) [inline onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.h:8]: Consider reorder your includes - https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: onnxruntime/core/providers/xnnpack/nn/conv.cc [hunk 5]
```diff
@@ -505,28 +505,29 @@ Status Conv::Compute(OpKernelContext* context) const {
   if (Y->Shape().Size() == 0) {
     return Status::OK();
   }
+  pthreadpool_t t_pool = GetThreadPool();
```
[the remark this answers is not in the document]
Comment (wejoncy) [reply onnxruntime/core/providers/xnnpack/nn/conv.cc:508]: `GetThreadPool()` is inline by default, we can prove it from its assembly code.

File: onnxruntime/core/session/provider_registration.cc (hunk 2 above)
Comment (wejoncy) [reply onnxruntime/core/session/provider_registration.cc:86]: Agree with pranavsharma, we should either implement all heavy ops or disable XNNPACK EP. Ort threadpool will also eats up half of CPUs even it's in IDLE with spinning. 
A workaround is to insert a spinning switch operation when switch from an EP to another EP, the operation should turn off the former's threadpool spinning and turn on the current threadpool spinning.

**Comment by wejoncy** [comment]
>

File: onnxruntime/core/providers/xnnpack/nn/softmax.h (hunk 3 above)
Comment (wejoncy) [reply onnxruntime/core/providers/xnnpack/nn/softmax.h:4]: Seems all files in ORT have similiar declaration headers, Do you mean we don't need these headers in `.h` file?

**Comment by wejoncy** [comment]
> The approach stratifies the threading between core ort and xnn EP - that leaves us a solid question, what if there are cpu ops expecting multi-threading? The choice we made here indeed brings implementation simplicity, while actually we could go extra steps to make pthread pool also visible to cpu ops to maintain parallelism. So - did we covered perf test on this case? Mind share the numbers as part of description?

There would be definitely models are running faster on CPUs as XNNPACK takes Conv2D as the main heavy ops and accelerate it. If a model contains Conv3D or other ops, which will expect multi-threading for cpu ops. But in that case, XNNPACK EP is not the priority.
xnn ep mainly target for conv based model and should be compatible for almost all heavy computation ops.
To allow pthreadpool to be xnnpack's private threadpool now is that fastest way to make multi-thread workable in xnn EP.

File: onnxruntime/core/session/provider_registration.cc (hunk 2 above)
Comment (skottmckay) [reply onnxruntime/core/session/provider_registration.cc:86]: Let's assume for XNNPACK to be a viable option it has kernels for all heavy ops. User can enable/disable based on perf testing.

I not sure we want to set any defaults for now as I don't think we know enough to know what is best in general, as that might be heavily dependent on the model and device. 

We should clearly document that an XNNPACK EP user should try two things. 
1) Set SessionOptions intra_op_num_threads to 1 and the xnnpack threadpool size to the number of cores (or whatever value they would have used for intra_op_num_threads if xnnpack wasn't enabled)
2) Set xnnpack threadpool size to the same size as intra_op_num_threads and disable spinning.

If we want to be really pro-active we could check if they did either of these things, and output a warning telling them to do so (message should point to doco so they understand the pros/cons of the two choice). 

The check might be something like there's no provider option for xnnpack EP setting number of threads, or ORT intra_op_num_threads > 1 and the disable spinning config option isn't set, output warning message.

File: onnxruntime/core/session/provider_registration.cc [hunk 6]
```diff
@@ -74,7 +75,15 @@ ORT_API_STATUS_IMPL(OrtApis::SessionOptionsAppendExecutionProvider,
 #endif
   } else if (strcmp(provider_name, "XNNPACK") == 0) {
 #if defined(USE_XNNPACK)
+    // If Xnnpack is enabled and user don't give a specific value, we need to set it same as cpu threadpool size
```
Comment (skottmckay) [inline onnxruntime/core/session/provider_registration.cc:78]: Need to update the documentation in onnxruntime_c_api.h that has info on the provider options for EPs that can be enabled via SessionOptionsAppendExecutionProvider.

Also need to add something equivalent to https://onnxruntime.ai/docs/execution-providers/SNPE-ExecutionProvider.html for xnnpack on the gh-pages branch now that the XNNPACK EP has a decent number of kernels and it's meaningful for users to try it out.

**Comment by wejoncy** [comment]
@RandySheriffH @yufenglee  this commit implemented "A unite threadpool" based on Pthreadpool. https://github.com/microsoft/onnxruntime/pull/11762/commits/352d17341165d3e11491e8b41ea03cc849edfa35

File: onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.h [hunk 7]
```diff
@@ -8,17 +8,23 @@
...
 namespace onnxruntime {
 // placeholder for future use. no options currently
 struct XnnpackExecutionProviderInfo {
+  int xnn_thread_pool_size{0};
   XnnpackExecutionProviderInfo() = default;
 
-  XnnpackExecutionProviderInfo(const ProviderOptions&) {
+  XnnpackExecutionProviderInfo(const ProviderOptions& po) {
+    if (po.count("thread_num")) {
+      xnn_thread_pool_size = std::stoi(po.at("thread_num"));
```
[the remark this answers is not in the document]
Comment (wejoncy) [reply onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.h]: Good suggestion.

I have to document it detail in onnxruntime.ai/docs which you paste below. I am not sure if we should update docs in this branch https://github.com/microsoft/onnxruntime/tree/gh-pages

File: onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.h [hunk 8]
```diff
@@ -38,6 +44,15 @@ class XnnpackExecutionProvider : public IExecutionProvider {
...
   bool ConcurrentRunSupported() const override { return false; }
+
+  pthreadpool_t GetPrivateThreadPool() const {
+    return xnnpack_thread_pool_.get();
+  }
+
+ private:
+  // Thread pool with smart-pointer for lifetime management.
+  std::unique_ptr<pthreadpool, decltype(&pthreadpool_destroy)> xnnpack_thread_pool_{
+      nullptr, &pthreadpool_destroy};
```
[the remark this answers is not in the document]
Comment (wejoncy) [reply onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.h:54]: If we use a default ctor. we have to initialize it like
```
XnnpackExecutionProvider::XnnpackExecutionProvider(const XnnpackExecutionProviderInfo& info)
    : IExecutionProvider{kXnnpackExecutionProvider, true},
      xnnpack_thread_pool_(std::unique_ptr<pthreadpool, decltype(&pthreadpool_destroy)>(nullptr, pthreadpool_destroy)) {
  if (info.xnn_thread_pool_size > 1) {
    xnnpack_thread_pool_.reset(
        pthreadpool_create(static_cast<size_t>(info.xnn_thread_pool_size)));
  }
}
```
So,  an explicit initialization here seems better.

Of course, we can get rid of unique_ptr and manage threadpool mannully.

File: onnxruntime/core/session/provider_registration.cc (hunk 2 above)
Comment (wejoncy) [reply onnxruntime/core/session/provider_registration.cc]: Either way, spinning should be disabled for better performance and lower power consumption.
We could warn something where config is not correct or not appropriate.
But we could do something like "always disable spinning", is that right?

File: onnxruntime/core/session/provider_registration.cc (hunk 6 above)
Comment (wejoncy) [reply onnxruntime/core/session/provider_registration.cc]: Will create another PR to update documents

File: include/onnxruntime/core/providers/providers.h [hunk 9]
```diff
@@ -5,9 +5,11 @@
...
 namespace onnxruntime {
 class IExecutionProvider;
+struct SessionOptions;
 
 struct IExecutionProviderFactory {
   virtual ~IExecutionProviderFactory() = default;
-  virtual std::unique_ptr<IExecutionProvider> CreateProvider() = 0;
+  virtual std::unique_ptr<IExecutionProvider> CreateProvider(
+      const SessionOptions* options = nullptr) = 0;
 };
```
Comment (skottmckay) [inline include/onnxruntime/core/providers/providers.h:14]: What's the reason for this change?

Comment (wejoncy) [reply include/onnxruntime/core/providers/providers.h:14]: Currently, How Ep created is decoupled with `SessionOptions`, so we don't know anything about what's the related configuration value when an EP was created.
However, for some EPs, they need to know if some configs were conflict with its unique features. Like CUDA/DML, that's why they have to hardcode into `inference_session.cc`.

For XNNPACK, we want to check if ort's intra-threadpool spinning and thread-pool size were compatible with XNNPACK EP, would it be best fit to check it when creating a EP?

Comment (skottmckay) [reply include/onnxruntime/core/providers/providers.h:14]: Given where we are in the release cycle, my suggestion would be to keep this PR minimal for now and not change the EP interface. We can document the suggested setup for xnnpack usage and also include that in the release notes so it has high visibility.

In the next release we can figure out what we want to do with SessionOptions and threadpool unification. 

@pranavsharma do you agree?

Comment (pranavsharma) [reply include/onnxruntime/core/providers/providers.h:14]: Makes sense. Let's defer the EP interface changes for now.

File: onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.h [hunk 10]
```diff
@@ -9,13 +9,17 @@
...
 struct XnnpackExecutionProviderInfo {
+  int xnn_thread_pool_size{0};
   XnnpackExecutionProviderInfo() = default;
 
-  XnnpackExecutionProviderInfo(const ProviderOptions&) {
-    // future: parse ProviderOptions
+  XnnpackExecutionProviderInfo(const ProviderOptions& po) {
+    if (po.count("intra_op_num_threads")) {
+      xnn_thread_pool_size = std::stoi(po.at("intra_op_num_threads"));
+    }
```
Comment (skottmckay) [inline onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.h:22]: nit: if you want to use the entry you're looking up use `find` so there's only one lookup (vs. one lookup for `count` and one for `at`).

**Review by skottmckay — APPROVED** [review]
:shipit:

File: onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.h (hunk 10 above)
Comment (wejoncy) [reply onnxruntime/core/providers/xnnpack/xnnpack_execution_provider.h:22]: Make sense with `if statement with initializer`

**Review by skottmckay — APPROVED** [review]
